frontend/projects/upgrade/src/app/features/dashboard/home/components/factorial-experiment-design/factorial-experiment-design.component.ts

```diff
@@ -0,0 +1,603 @@
+
+  emitEvent(eventType: NewExperimentDialogEvents) {
+    switch (eventType) {
+      case NewExperimentDialogEvents.CLOSE_DIALOG:
+        if (
+          this.factorialExperimentDesignForm.dirty ||
+          this.experimentDesignStepperService.getHasExperimentDesignStepperDataChanged()
+        ) {
+          this.dialogService
+            .openConfirmDialog()
+            .afterClosed()
+            .subscribe((res) => {
+              if (res) {
+                this.emitExperimentDialogEvent.emit({ type: eventType });
+              }
+            });
+        } else {
+          this.emitExperimentDialogEvent.emit({ type: eventType });
+        }
+        break;
+      case NewExperimentDialogEvents.SEND_FORM_DATA:
+        if (this.factorialExperimentDesignForm.dirty) {
+          this.experimentDesignStepperService.experimentStepperDataChanged();
+        }
+        if (!this.isExperimentEditable) {
+          this.emitExperimentDialogEvent.emit({
+            type: eventType,
+            formData: this.experimentInfo,
+            path: NewExperimentPaths.EXPERIMENT_DESIGN,
+          });
+          break;
+        }
+        this.saveData(eventType);
+        break;
+      case NewExperimentDialogEvents.SAVE_DATA:
+        if (!this.isExperimentEditable) {
+          this.emitExperimentDialogEvent.emit({
+            type: eventType,
+            formData: this.experimentInfo,
+            path: NewExperimentPaths.EXPERIMENT_DESIGN,
+          });
+          break;
+        }
+        this.saveData(eventType);
+        this.experimentDesignStepperService.experimentStepperDataReset();
+        this.isAnyRowRemoved = false;
+        this.factorialExperimentDesignForm.markAsPristine();
+        break;
+    }
+  }
+
+  saveData(eventType) {
+    this.validateForm();
+    this.validateFactors();
+
+    if (!this.conditionTableDataUpToDate) {
+      this.experimentDesignStepperService.updateFactorialDesignData(this.factorialExperimentDesignForm.value);
+    }
+
+    if (this.isFormValid()) {
+      const factorialExperimentDesignFormData = this.factorialExperimentDesignForm.value;
+      const factorialPartitions = this.convertToPartitionData(factorialExperimentDesignFormData);
+      const factorialConditions = this.experimentDesignStepperService.createFactorialConditionRequestObject();
+
+      const factorialConditionAliases: ExperimentConditionAliasRequestObject[] =
+        this.experimentDesignStepperService.createFactorialConditionsConditionAliasesRequestObject();
+
+      this.emitExperimentDialogEvent.emit({
+        type: eventType,
+        formData: {
+          conditions: factorialConditions,
+          partitions: factorialPartitions,
+          conditionAliases: factorialConditionAliases,
+        },
+        path: NewExperimentPaths.EXPERIMENT_DESIGN,
+      });
+
+      // scroll back to the factors table
+      this.scrollToFactorsTable();
+    }
+  }
+
+  convertToPartitionData(factorialExperimentDesignFormData) {
```

Comment:
@Yagnik56 @RidhamShah @ppratikcr7 @VivekFitkariwala This is the type of function I'm looking at when I say we need to clean things up, it's not really ready for PR. I'd like to prevent us from being okay with continuing to merge in code that works functionally but is a not cleaned up for the benefit of other developers on the team, it's really hard to work this way. 

- no types are provided
- outdated terms are used (partition)
- code complexity is too high for routine operations, this can easily be broken down into small readable functions

This kind of logic could probably be moved to stepper service also, that's what it's there for, as there's no dependency on the component class or template bindings.

Reply:
type was missed, it is added now

Reply:
ok, sorry to pick on this particular function, it's not at all the most complex one in here, just happened to be where I got frustrated trying to review because the PR has a lot of changes and it's not always easy to read.

I guess we have different philosophies on coding style, so maybe I'll just give this up if the rest of the team disagrees and does not find this PR hard to read, but I want to explain why I made this comment and we can move on.

The thing is I see lots of long functions with nested conditionals and subtasks in this PR (and in the codebase in general) that could easily be simplified, purely for the readability of future developers, and I think that is a hugely important thing.

When you're working with existing code, would you rather see small functions that can be quickly understood by a even a new developer, or dig through long mazes of nested conditionals and anonymous functions that you have to go find the original developer to explain?

Or look at it this way, there's another huge benefit: If I asked you to write a jest unit test for this function, how many cases would it take to get 100% coverage? How many ways would you have to mock up data? How about the `ngOnInit` function, which has all kinds of things going on? That's what I mean by code complexity. If it's hard for a machine to process all the possibilities, a human is going to make mistakes. A set of small pure functions is exponentially easier to mock and test than one umbrella function that technically accomplishes exactly the same tasks.

The hard work of getting the functionality is done, polishing the code for readability is free and easy. Once it gets merged in, we can't really say we'll clean it up later, because that rarely actually happens. Better spend a little extra time now than spend weeks debugging complex code later.

Reply:
and obviously, please let me know where my contributions to this PR can be improved also. I almost never hear from anyone on Playpower in PRs so I never know what others are thinking (until I've broken something in dev branch).

Reply:
@danoswaltCL You are absolutely right. We should do as much simplification as possible to make code readable friendly for all. We will make sure to get it done in this PR and future PR's as well. There is no significant improvements that I could see in your contributions, else would have definitely pointed it out. Just a few console logs that you have kept in the conditions table component. :)